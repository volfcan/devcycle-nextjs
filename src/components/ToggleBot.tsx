'use client'
import { useAllFeatures, useVariableValue } from '@devcycle/nextjs-sdk';
import classNames from 'classnames';
import Image from 'next/image';

/**
 * Example client component
 */
function ToggleBot() {
  /**
   * The useVariableValue hook will return the current value of a variable.
   * If no value is defined for the current user, the default value will be returned.
   */
  const shouldWink = useVariableValue('togglebot-wink', false)
  const spinSpeed = useVariableValue('togglebot-speed', 'off')

  const features = useAllFeatures()
  const { variationName = 'Default' } = features['hello-togglebot'] ?? {}

  return (
    <div className="App-content">
      <div className="ToggleBot-message" data-testid="togglebot-message">
        &quot;{getMessage(spinSpeed)}&quot;
      </div>
      <Image
        src={getImageSource(spinSpeed, shouldWink)}
        className={classNames(['ToggleBot-logo', `spin-${spinSpeed}`])}
        alt="togglebot"
        width={240}
        height={198}
      />
      <div className="ToggleBot-variation">
        Serving Variation: <b>&quot;{variationName}&quot;</b>
      </div>
    </div>
  );
}

const getMessage = (spinSpeed: string) => {
  switch (spinSpeed) {
    case 'slow':
      return 'Awesome, look at you go!'
    case 'fast':
      return 'This is fun!'
    case 'off-axis':
      return '...I\'m gonna be sick...'
    case 'surprise':
      return 'What the unicorn?'
    default:
      return 'Hello! Nice to meet you.'
  }
}

const getImageSource = (spinSpeed: string, shouldWink: boolean) => {
  if (spinSpeed === 'surprise') return 'unicorn.svg'
  return shouldWink ? '/togglebot-wink.svg' : '/togglebot.svg'
}

export default ToggleBot;
