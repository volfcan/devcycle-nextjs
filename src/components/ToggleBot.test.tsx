import { render, screen } from '@testing-library/react';
import * as dvc from '@devcycle/nextjs-sdk';

import ToggleBot from './ToggleBot';

jest.mock('@devcycle/nextjs-sdk')

describe('ToggleBot', () => {
  let mockUseVariableValue: jest.SpyInstance
  let mockUseAllFeatures: jest.SpyInstance

  beforeEach(() => {
    mockUseVariableValue = jest.spyOn(dvc, 'useVariableValue')
    mockUseAllFeatures = jest.spyOn(dvc, 'useAllFeatures')
  })

  test.each([
    'off',
    'slow',
    'fast',
    'off-axis',
    'surprise'
  ])('renders message text for speed "%s"', (speed) => {
    mockUseVariableValue.mockImplementation((key, defaultValue) => {
      if (key === 'togglebot-speed') return speed
      return defaultValue
    })
    mockUseAllFeatures.mockReturnValue({
      'hello-togglebot': {
        variationName: 'Default'
      }
    })

    render(<ToggleBot />);
    const element = screen.getByTestId('togglebot-message')
    expect(element.textContent).toMatchSnapshot()
  })
})
