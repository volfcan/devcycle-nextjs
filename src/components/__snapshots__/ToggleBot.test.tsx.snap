// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleBot renders message text for speed "fast" 1`] = `""This is fun!""`;

exports[`ToggleBot renders message text for speed "off" 1`] = `""Hello! Nice to meet you.""`;

exports[`ToggleBot renders message text for speed "off-axis" 1`] = `""...I'm gonna be sick...""`;

exports[`ToggleBot renders message text for speed "slow" 1`] = `""Awesome, look at you go!""`;

exports[`ToggleBot renders message text for speed "surprise" 1`] = `""What the unicorn?""`;
